### ISSUE TITLE
Automated upgrade fails to upgrade v1.18.3+k3s1 to stable

### ISSUE THREAD

**Issue by gizmotronic** [rancher/k3s#1954]
**Version:**
k3s version v1.18.3+k3s1 (96653e8d)

**K3s arguments:**
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--no-deploy servicelb,local-storage,traefik" sh -

**Describe the bug**
When I configure automated upgrades using a release channel, nothing happens.

**To Reproduce**
Configure automated upgrades exactly as indicated on the [Automated Upgrades](https://rancher.com/docs/k3s/latest/en/upgrades/automated/) page, using `channel: https://update.k3s.io/v1-release/channels/stable` instead of `version:`.

**Expected behavior**
I expect the cluster to be upgraded to v1.18.4+k3s1.

**Actual behavior**
Upgrade jobs are not created and the cluster is not upgraded. I find repeating messages in the system-upgrade-controller log for the k3s server and one of the worker nodes that indicate that the upgrade jobs can't be created due to a problem with the name.

**Workaround**
This problem seems to affect only upgrades using release channels. I most recently upgraded to v1.18.3+k3s1 without any issue by using `version:` instead of `channel:`.

**Additional context / logs**

```
E0625 14:35:09.072223       1 controller.go:135] error syncing 'system-upgrade/server-plan': handler system-upgrade-controller: Get https://update.k3s.io/v1-release/channels/stable: x509: certificate is valid for [redacted], not update.k3s.io, handler system-upgrade-controller: failed to create system-upgrade/apply-server-plan-on-tv-land-with- batch/v1, Kind=Job for system-upgrade-controller system-upgrade/server-plan: Job.batch "apply-server-plan-on-tv-land-with-" is invalid: [metadata.name: Invalid value: "apply-server-plan-on-tv-land-with-": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.template.labels: Invalid value: "apply-server-plan-on-tv-land-with-": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')], requeuing
E0625 14:35:13.208914       1 controller.go:135] error syncing 'system-upgrade/agent-plan': handler system-upgrade-controller: Get https://update.k3s.io/v1-release/channels/stable: x509: certificate is valid for [redacted], not update.k3s.io, handler system-upgrade-controller: failed to create system-upgrade/apply-agent-plan-on-raspberrypi-with- batch/v1, Kind=Job for system-upgrade-controller system-upgrade/agent-plan: Job.batch "apply-agent-plan-on-raspberrypi-with-" is invalid: [metadata.name: Invalid value: "apply-agent-plan-on-raspberrypi-with-": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.template.labels: Invalid value: "apply-agent-plan-on-raspberrypi-with-": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')], requeuing
```

**Comment by ShylajaDevadiga** [comment]
Could you provide the version of System Upgrade Controller used and the plan yaml for us to repro this issue?

**Comment by dweomer** [comment]
@gizmotronic the root of the problem is indicated by the first part of the supplied error messages:
```
E0625 14:35:09.072223       1 controller.go:135] error syncing 'system-upgrade/server-plan': handler system-upgrade-controller: Get https://update.k3s.io/v1-release/channels/stable: x509: certificate is valid for [redacted], not update.k3s.io, ...
```

More specifically: **x509: certificate is valid for [redacted], not update.k3s.io**. What is curious here is that you seem to have redacted the resolved host. Is this because it is a host you control? If so you might be better served by using the valid host/domain in the channel URL.

Alternatively, depending on your network setup, it is possible that your DNS is giving you bad results. What does `update.k3s.io` resolve to on the node that the controller is running on? For that matter, what does `curl https://update.k3s.io/v1-release/channels/stable -D-`. My local results (k3OS v0.10.3 and Ubuntu 20.0.4):
```
nuc-1 [~]$ nslookup update.k3s.io
Server:		192.168.1.1
Address:	192.168.1.1:53

Non-authoritative answer:
update.k3s.io	canonical name = ext-services-lb-2a685576fdf664e2.elb.us-west-2.amazonaws.com
Name:	ext-services-lb-2a685576fdf664e2.elb.us-west-2.amazonaws.com
Address: 52.36.54.134
Name:	ext-services-lb-2a685576fdf664e2.elb.us-west-2.amazonaws.com
Address: 34.208.213.149

Non-authoritative answer:
update.k3s.io	canonical name = ext-services-lb-2a685576fdf664e2.elb.us-west-2.amazonaws.com

nuc-1 [~]$ curl https://update.k3s.io/v1-release/channels/stable -D-
HTTP/2 302 
server: openresty/1.15.8.1
date: Thu, 25 Jun 2020 19:03:54 GMT
content-type: text/html; charset=utf-8
content-length: 79
location: https://github.com/rancher/k3s/releases/tag/v1.18.4+k3s1
strict-transport-security: max-age=15724800; includeSubDomains

<a href="https://github.com/rancher/k3s/releases/tag/v1.18.4+k3s1">Found</a>.

nuc-1 [~]$ 
```

**Comment by brandond** [comment]
@dweomer beat me to it while I was typing - is there any reason in particular that you redacted the hostname from the TLS error? This suggests that you have something in your environment that is interfering with the connection to the update service - perhaps DNS redirection, a proxy server, or firewall?

`Get https://update.k3s.io/v1-release/channels/stable: x509: certificate is valid for [redacted], not update.k3s.io`

The failure to pull the release from the channel server somehow ended up not being a terminal failure, which it should have been. Instead it resulted in the plan status not populating properly, so the LatestHash is an empty string at https://github.com/rancher/system-upgrade-controller/blob/master/pkg/upgrade/job/job.go#L98 which results in generation of an invalid job name.

`Job.batch "apply-server-plan-on-tv-land-with-" is invalid`

**Comment by gizmotronic** [comment]
> [@ShylajaDevadiga] Could you provide the version of System Upgrade Controller used and the plan yaml for us to repro this issue?

I'm using rancher/system-upgrade-controller:v0.4.0 as is indicated on the Automated Upgrades page that I originally linked to. I copied the plans from that same page, removed `version:`, and added `channel:`.

> [@dweomer] More specifically: x509: certificate is valid for [redacted], not update.k3s.io. What is curious here is that you seem to have redacted the resolved host. Is this because it is a host you control? If so you might be better served by using the valid host/domain in the channel URL.

I've actually redacted the SAN names from my domain's certificate, not the host that was resolved. I avoided publishing my domain details for spammers to scrape. Replacing it with `example.com`, that part of the message would render thus:

```
handler system-upgrade-controller: Get https://update.k3s.io/v1-release/channels/stable: x509: certificate is valid for example.com, *.example.com, *.subdomain.example.com, not update.k3s.io
```

This is important, though, and I'll get to that in a moment.

> Alternatively, depending on your network setup, it is possible that your DNS is giving you bad results. What does update.k3s.io resolve to on the node that the controller is running on? For that matter, what does curl https://update.k3s.io/v1-release/channels/stable -D-.

The node the controller is running on, tv-land, is an Ubuntu 18.04 system. Here's the name resolution:

```
$ host update.k3s.io
update.k3s.io is an alias for ext-services-lb-2a685576fdf664e2.elb.us-west-2.amazonaws.com.
ext-services-lb-2a685576fdf664e2.elb.us-west-2.amazonaws.com has address 52.36.54.134
ext-services-lb-2a685576fdf664e2.elb.us-west-2.amazonaws.com has address 34.208.213.149
```

And,

```
$ curl https://update.k3s.io/v1-release/channels/stable -D-
HTTP/2 302 
server: openresty/1.15.8.1
date: Thu, 25 Jun 2020 21:59:44 GMT
content-type: text/html; charset=utf-8
content-length: 79
location: https://github.com/rancher/k3s/releases/tag/v1.18.4+k3s1
strict-transport-security: max-age=15724800; includeSubDomains

<a href="https://github.com/rancher/k3s/releases/tag/v1.18.4+k3s1">Found</a>.

```

I appreciate that the problem is something local to my setup, and that I've uncovered a different problem than what I originally expected (thanks for the explanation, @brandond).

My firewall is a pfSense setup that should be allowing outbound traffic through unimpeded. There's no outbound proxy or DNS redirection going on. pfSense also provides DNS service on my network using [Unbound](https://nlnetlabs.nl/projects/unbound/about/), which generally works quite well. Having said that, though, the problem bears some similarity to an issue I had with the cert-manager Helm chart on a Rancher-launched cluster a while back. The only solution in that case was to configure the pod to use OpenDNS directly (for the DNS01 challenge provider) rather than relying on the node's DNS configuration. If I recall correctly this was because Unbound doesn't work as an authoritative name server.

Based on your feedback I got a shell into another container running on the same node. The curl command failed with a similar message about the certificate. Then, when I tried to update the package list to install some bits to help debug things, I got back a 404 error listing my external IP address when trying to load Release files from http://archive.ubuntu.com/ubuntu.

I'll keep digging.

**Comment by gizmotronic** [comment]
This turns out to be a DNS problem that only manifested in by accident. It's related to #53 and very similar to #999. As with the latter issue I have a wildcard DNS configuration that redirects (anything).example.com to my reverse proxy.

My k3s server is running on an amd64 machine running Ubuntu Server 18.04. The two worker nodes were running Raspbian Buster. The cluster ran numerous versions up to v1.17.4+k3s1 without any DNS problems manifesting.

I recently upgraded the cluster to v1.18.3+k3s1. I hadn't noticed any problems with any other workloads running on the Ubuntu node. Workloads running on the Raspbian node definitely didn't have any problem. It's worth noting that all of the existing workloads were using host networking and dnsPolicy was set to `ClusterFirstWithHostNet`.

As an attempt to resolve the issue I updated system-upgrade-controller use host networking and configured dnsPolicy. This didn't change anything. However, I was able to reproduce the problem by running an Ubuntu pod on the same node. It didn't matter whether I used host networking or not. The Raspbian nodes continued to work when using the host network, but an Ubuntu pod using cluster networking also had trouble.

Finally, I upgraded the Raspbian nodes to Ubuntu Server 20.04 (aarch64). This caused all DNS lookups to resolve to my external IP address, as has been happening all along with the k3s server on Ubuntu. I thought this might be related to systemd-resolved. Disabling it didn't change anything.

As discussed in #53, I'm able to work around the issue by updating dnsConfig to set `ndots` to 1. This prevents the pod DNS lookup from appending local search paths and getting a false positive result from my DNS server.

I'm closing the ticket since this is clearly not a problem with automated upgrades. Thanks to everyone who took the time to respond.
